src/cast_core/src/StreamStatsReporter.cpp

```diff
@@ -0,0 +1,113 @@
+// Copyright 2023-present MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cast_core/actors/StreamStatsReporter.hpp>
+
+#include <memory>
+
+#include <yaml-cpp/yaml.h>
+
+#include <bsoncxx/array/view.hpp>
+#include <bsoncxx/builder/stream/document.hpp>
+#include <bsoncxx/json.hpp>
+
+#include <mongocxx/client.hpp>
+#include <mongocxx/collection.hpp>
+#include <mongocxx/database.hpp>
+
+#include <boost/log/trivial.hpp>
+#include <boost/throw_exception.hpp>
+
+#include <gennylib/Cast.hpp>
+#include <gennylib/MongoException.hpp>
+#include <gennylib/context.hpp>
+
+#include <value_generators/DocumentGenerator.hpp>
+
+
+namespace genny::actor {
+
+struct StreamStatsReporter::PhaseConfig {
+    mongocxx::database database;
+    TimeSpec pollInterval;
+    bsoncxx::document::value metricsCommand = bsoncxx::from_json("{}");
+
+    PhaseConfig(PhaseContext& phaseContext, mongocxx::database&& db, ActorId id)
+        : database{db},
+          pollInterval{phaseContext["PollInterval"].to<TimeSpec>()} {
+        int threads = phaseContext.actor()["Threads"].to<int>();
+        assert(threads == 1);
+
+        std::string streamProcessorName = phaseContext["StreamProcessorName"].to<std::string>();
+        bsoncxx::builder::stream::document builder;
+        metricsCommand = builder << "streams_getMetrics" << "" << bsoncxx::builder::stream::finalize;
+    }
+};
+
+void StreamStatsReporter::run() {
+    for (auto&& config : _loop) {
+        for (const auto&& _ : config) {
+            auto throughput = _throughput.start();
+            std::this_thread::sleep_for(genny::Duration(config->pollInterval));
+
+            auto res = config->database.run_command(config->metricsCommand.view());
+            BOOST_LOG_TRIVIAL(debug) << "metrics reply: " << bsoncxx::to_json(res);
+            bsoncxx::array::view counters = res.view()["counters"].get_array();
+
+            bool inputDocuments{false};
+            bool inputBytes{false};
+
+            for (auto c : counters) {
+                auto name = c["name"].get_string().value;
+                int64_t value = c["value"].get_int64();
+
+                if (name == "num_input_documents") {
+                    assert(!inputDocuments);
```

Comment:
Instead of `assert` there's `BOOST_THROW_EXCEPTION` that will give us nice stackframes.